art/estimators/certification/randomized_smoothing/smooth_mix/pytorch.py

```diff
@@ -0,0 +1,337 @@
+            warmup_v = min(1.0, (epoch + 1) / self.warmup)
+
+            for x_batch, y_batch in dataloader:
+                mini_batch_size = len(x_batch) // self.num_noise_vec
+
+                for mini_batch in range(self.num_noise_vec):
+                    # Create mini batch
+                    inputs = x_batch[mini_batch * mini_batch_size : (mini_batch + 1) * mini_batch_size]
+                    labels = y_batch[mini_batch * mini_batch_size : (mini_batch + 1) * mini_batch_size]
+
+                    # Move inputs to GPU
+                    inputs = inputs.to(self.device)
+                    labels = labels.to(self.device)
+
+                    noises = [torch.randn_like(inputs) * self.scale for _ in range(self.num_noise_vec)]
+
+                    # Attack and find adversarial examples
+                    self._model.eval()
+                    inputs, inputs_adv = self._smooth_mix_pgd_attack(inputs, labels, noises, warmup_v)
+                    self._model.train(mode=training_mode)
+
+                    in_clean_c = torch.cat([inputs + noise for noise in noises], dim=0)
+                    logits_c = self._model(in_clean_c)[-1]
+                    labels_c = labels.repeat(self.num_noise_vec)
+
+                    logits_c_chunk = torch.chunk(logits_c, self.num_noise_vec, dim=0)
+                    clean_sm = F.softmax(torch.stack(logits_c_chunk), dim=-1)
+                    clean_avg_sm = torch.mean(clean_sm, dim=0)
+                    loss_xent = F.cross_entropy(logits_c, labels_c, reduction="none")
+
+                    # mix adversarial examples
+                    in_mix, labels_mix = self._mix_data(inputs, inputs_adv, clean_avg_sm)
+                    in_mix_c = torch.cat([in_mix + noise for noise in noises], dim=0)
+                    labels_mix_c = labels_mix.repeat(self.num_noise_vec, 1)
+                    logits_mix_c = F.log_softmax(self._model(in_mix_c)[-1], dim=1)
+
+                    preds = torch.argmax(clean_avg_sm, dim=-1)
+                    ind_correct = (preds == labels).float()
+                    ind_correct = ind_correct.repeat(self.num_noise_vec)
+
+                    loss_mixup = F.kl_div(logits_mix_c, labels_mix_c, reduction="none").sum(1)
+                    loss = loss_xent.mean() + self.eta * warmup_v * (ind_correct * loss_mixup).mean()
+
+                    # compute gradient and do SGD step
+                    self._optimizer.zero_grad()
+                    loss.backward()
+                    self._optimizer.step()
+
+            if scheduler is not None:
+                scheduler.step()
+
+    def _smooth_mix_pgd_attack(
+        self,
+        inputs: "torch.Tensor",
+        labels: "torch.Tensor",
+        noises: List["torch.Tensor"],
+        warmup_v: float,
+    ) -> Tuple["torch.Tensor", "torch.Tensor"]:
+        """
+        The authors' implementation of the SmoothMixPGD attack.
+        Code modified from https://github.com/jh-jeong/smoothmix/code/train.py
```

Comment:
Let's include their MIT License text (https://github.com/jh-jeong/smoothmix/blob/main/LICENSE) at the top of our new file after ART's MIT license text.
```suggestion
        Code modified from https://github.com/jh-jeong/smoothmix/blob/main/code/train.py
```